internal/nozzle/nozzle.go

```diff
@@ -112,63 +113,74 @@ func (d *Nozzle) Start() error {
 		d.log,
 		d.db)
 
-	// Let's store the last three times when a restart occurs, so that we don't keep trying forever
-	var retryHistory = make([]time.Time, 3)
-	for {
-		// consume messages from the Firehose and push them to d.messages
-		d.consumer, err = d.consumeFirehose(authToken)
-		if err != nil {
-			return err
-		}
-		d.messages, d.errs = d.consumer.FilteredFirehose(d.config.FirehoseSubscriptionID, authToken, consumer.Metrics)
+	// Initialize the firehose consumer (with retry enable)
+	err = d.startFirehoseConsumer(authToken)
+	if err != nil {
+		return err
+	}
 
-		// Start multiple workers to parallelize firehose events (event.envelope) transformation into processedMetrics
-		// and then grouped into metricsMap
-		d.startWorkers()
-		// Start infinite loop to periodically submit metrics (metricsMap) to Datadog
-		err = d.postToDatadog()
+	// Start multiple workers to parallelize firehose events (event.envelope) transformation into processedMetrics
+	// and then grouped into metricsMap
+	d.startWorkers()
 
-		if !shouldReconnect(err) {
-			break
-		}
+	// Execute infinite loop.
+	// This method is blocking until we get error or a stop signal
+	err = d.Run()
 
-		// memorize the retry timestamp
-		retryHistory = append(retryHistory[1:], time.Now())
-		if retryHistory[2].Sub(retryHistory[0]) < 5*time.Second {
-			// We retried three times quickly, there might be a larger issue
-			// Let's break out of the loop and return the error
-			d.log.Info("Too many retries, shutting down...")
-			break
-		}
-		d.stopWorkers()
-		// Sleep a little to not reconnect to quickly
-		time.Sleep(500 * time.Millisecond)
-		d.log.Info("Websocket connection lost, reestablishing connection...")
-	}
-
-	// Whenever a stop signal is received the infinite loop within postToDatadog is stopped and then we stop all workers
+	// Whenever a stop signal is received the Run methode above will return. The code below will then be executed
 	d.log.Info("DataDog Firehose Nozzle shutting down...")
+	// Close Firehose Consumer
+	d.log.Infof("Closing connection with traffic controller due to %v", err)
+	d.consumer.Close()
+	// stop processor
 	d.stopWorkers()
+	// Submit metrics left in cache if any
+	d.PostMetrics()
 
 	return err
 }
 
-// Define the error cases where we should reattempt a connection
-func shouldReconnect(err error) bool {
-	if err == nil {
-		return false
+func (n *Nozzle) startFirehoseConsumer(authToken string) error {
```

Comment:
Can we name the object consistently ? Either `n` or `d` everywhere. (I prefer `n` :) )